providers/dell/idrac9/idrac.go

```diff
@@ -0,0 +1,229 @@
+package idrac9
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"strconv"
+
+	"github.com/bmc-toolbox/bmclib/providers"
+	"github.com/go-logr/logr"
+	"github.com/jacobweinstock/registrar"
+	"github.com/pkg/errors"
+)
+
+// Conn for dell idrac9 connections
+type Conn struct {
+	Host      string
+	Port      string
+	User      string
+	Pass      string
+	Log       logr.Logger
+	conn      *http.Client
+	xsrfToken string
+}
+
+const (
+	// ProviderName for the provider implementation
+	ProviderName = "idrac9"
+	// ProviderProtocol for the provider implementation
+	ProviderProtocol = "webgui"
+)
+
+var (
+	// Features implemented by dell idrac9 provider
+	Features = registrar.Features{
+		providers.FeatureUserCreate,
+		providers.FeatureUserUpdate,
+		providers.FeatureUserRead,
+		providers.FeatureUserDelete,
+	}
+)
+
+func (c *Conn) Open(ctx context.Context) error {
+	idrac := &IDrac9{ip: c.Host, username: c.User, password: c.Pass, log: c.Log}
+	err := idrac.httpLogin()
+	if err != nil {
+		return err
+	}
+	if idrac.httpClient == nil {
+		return errors.New("error opening connection")
+	}
+	c.conn = idrac.httpClient
+	c.xsrfToken = idrac.xsrfToken
+	return nil
+}
+
+func (c *Conn) Close(ctx context.Context) error {
+	idrac := &IDrac9{ip: c.Host, username: c.User, password: c.Pass, log: c.Log}
+	idrac.xsrfToken = c.xsrfToken
+	idrac.httpClient = c.conn
+	if idrac.httpClient != nil {
+		if _, _, err := idrac.delete("sysmgmt/2015/bmc/session"); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// Compatible tests whether a BMC is compatible with the idrac9 provider
+func (c *Conn) Compatible(ctx context.Context) bool {
+	err := c.Open(ctx)
```

Comment:
Hmm...for me a failed auth is the same as not being an idrac9 device or the same as the bmc being down or the IP of the bmc not being correct. given a `bmclib.Client` with specific IP, user, and pass defined any of the above scenarios constitute incompatibility. 

What are the consequences of attempting an idrac9 web GUI auth on a non-idrac9 bmc?

Reply:
> a failed auth is the same as not being an iDrac9 device or the same as the bmc being down or the IP of the bmc not being correct.

Being able to differentiate an auth failure `401`/ `403` from a failed request (IP unreachable/404s)
has benefits for the client - which enables it to,
- retry with a different password if necessary
- backoff if the iDrac9 has banned the client's IP or has blocked it because reached its connection/session limit

I see value in not defining a vendor specific compatibility method for generic drivers like IPMI/Redfish,
although, for provider specific drivers like this one, a vendor specific compatibility method would reduce the number of requests made to the BMC  (this would only matter over time, when BMCs have been running for a while without a reset)

Since a `401`/`403` clearly indicates an auth failure the [Opener](https://github.com/bmc-toolbox/bmclib/blob/13d83dfbc10ecf14da33837a6b2a3801a2eaa81b/bmc/connection.go#L52) could just return instead of trying another driver.

Reply:
Apologies, I don't know if I'm following here. are you saying that this existing idrac9 compatibility function should differentiate between auth failure from other failures? or are you saying that the compatibility check should make the user aware of the failure type? or that we should limit the number of request to a bmc as much as possible? or something else?

_"Since a 401/403 clearly indicates an auth failure the Opener could just return instead of trying another driver."_
I'm not sure exactly what the suggestion here is? This is instead of a compatibility check?

Reply:
Apologies, I was trying to convey - scenarios where differentiating between auth and other failures will be useful to the client, along with the points below,

a. The Compatibility method to return bool based on if the device was actually identified as an iDrac9 (for reasons in the previous comment) 
b. bmclib to keep BMC requests as relevant/minimal as possible (point a. enables this)
c. `Open` to return an Auth error for failed auth, along with the compatibility check (as a ticket for later)

Reply:
thank you for clarifying. follow up question. how do you define identified/compatible? I think this might be where we aren't understanding each other.

Reply:
> how do you define identified/compatible?

If the vendor specific driver queries a BMC and receives an expected response which is unique to the BMC, it could be considered to be identified and compatible, although an unexpected response (specifically `401`, `403`) should not mean the driver is unidentified or in-compatible. 

related, we're invoking `Open` in the Compatible method and not reusing the session either.